glib/src/subclass/shared.rs

```diff
@@ -0,0 +1,125 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+//! Module for registering shared types for Rust types.
+
+use crate::translate::*;
+
+/// Trait for defining shared types.
+///
+/// Links together the type name with the type itself.
+///
+/// See [`register_shared_type`] for registering an implementation of this trait
+/// with the type system.
+///
+/// [`register_shared_type`]: fn.register_shared_type.html
+pub trait SharedType: Clone + Sized + 'static {
+    /// Shared type name.
+    ///
+    /// This must be unique in the whole process.
+    const NAME: &'static str;
+
+    /// The inner refcounted type
+    type InnerType;
```

Comment:
With that separate trait you'd then here require the inner type to implement that, and would move the `ref_` / `unref` functions into that separate trait

Reply:
Also I think for the whole thing to work in a safe way, you need a `fn from_inner(Self::InnerType) -> Self` and `fn into_inner(Self) -> Self::InnerType` or similar